mediatimestamp/immutable/timerange.py

```diff
@@ -817,38 +817,26 @@ def into_chunks(
         :param time_duration: A TimeOffset representing the requested chunk length
         :returns: A TimeRange generator that generates chunks.
         """
-
-        last_tr = None
-        if self.start is None:
-            raise ValueError("This TimeRange object has no 'start' value and so cannot be chunked!")
-        remaining_length: int = cast(TimeOffset, self.length).to_nanosec() - time_duration.to_nanosec()
-        # If the input time range never ends we simply have a never-ending generator -- if self.end is None this
-        # will run forever.
-        while self.end is None or remaining_length > 0:
-            if last_tr is None:  # This is the first and not the last
-                # The first must always have the same start inclusivity as the timerange, and an exclusive ) end.
-                # The four possibilities for inclusivity are defined in this file as binary integers 0-3.
-                # The first bit is the end inclusivity - for exclusive ends the first bit must be zero, therefore
-                # we only need to use the last bit which indicates the start inclusivity.
-                last_tr = self.from_start_length(self.start, time_duration, self.Inclusivity(self.includes_start()))
-            else:  # this is not the first and not the last; it must have inclusive start and exclusive end
-                # We can also continue on from the end of the last timerange for our start timestamp.
-                if last_tr.end is not None:
-                    last_tr = self.from_start_length(last_tr.end, time_duration, self.INCLUDE_START)
-                else:
-                    raise ValueError("Something has gone badly wrong, the last chunk had no end value set!")
-            remaining_length -= time_duration.to_nanosec()
-            yield last_tr
-
-        # The final chunk.
-        # The sum of the requested duration and time remaining will determine the length of the final chunk.
-        if last_tr is None:  # There is only one chunk, and it's an exact fit OR shorter than the requested dur.
-            yield self.from_timerange(self)
-        else:
-            if last_tr.end is not None:
-                yield self.from_start_length(
-                    last_tr.end, self.end - last_tr.end,
-                    self.Inclusivity((self.includes_end() << 1) + 1))
-            else:
-                raise ValueError("Something has gone badly wrong, the last chunk had no end value set!")
+        remainder = self  # this is a timerange from the end of the last one to the end of self
+        while not remainder.is_empty():
+            next_timerange = TimeRange.from_start_length(
+                remainder.start,
+                time_duration,
+                self._get_self_inclusivity(remainder, time_duration))
```

Comment:
A couple things that I think could be improved here. The name `_get_self_inclusivity` and it being a member method doesn't look right for something that is specific to this into_chunks method. I would've moved this function into into_chunks. However, I'm not sure a function is really needed.

The start inclusivity case is taken care of by the intersect_with which will demote it to not INCLUDE_START if self doesn't.

The end inclusivity case can be a relatively simple if statement and assignment, e.g.
```
if next_timerange.end == remainder.end and remainder.includes_end():
  next_timerange.inclusivity |= Timerange.INCLUDE_END
```

Reply:
What about in the first run of the loop when no `next_timerange` exists? This is why I created the function -- I couldn't think of a neater way around it.

Reply:
Good point! I'd forgotten that this TimeRange class is immutable. You could do this instead:
```
if next_timerange.end == remainder.end and remainder.includes_end():
  next_timerange = TimeRange(
    next_timerange.start,
    next_timerange.end,
    next_timerange.inclusivity | TimeRange.INCLUDE_END
  )
```
I think on balance it is better than the separate function as it's more obvious what it is doing.

Reply:
Timeranges are immutable so I don't think it's possible to set the inclusivity _after_ instantiation.

Reply:
Got around this by creating a new next_timerange